FF1Lib/OverworldMap.cs

```diff
@@ -1690,61 +1696,119 @@ public void ShuffleObjectiveNPCs(MT19337 rng)
+			new byte[] { OverworldTiles.ASTOS_CASTLE_W, OverworldTiles.ASTOS_CASTLE_E },
+			new byte[] { OverworldTiles.ICE_CAVE },
+			new byte[] { OverworldTiles.DWARF_CAVE },
+			new byte[] { OverworldTiles.MATOYAS_CAVE },
+			new byte[] { OverworldTiles.TITAN_CAVE_E, OverworldTiles.TITAN_CAVE_W },
+			new byte[] { OverworldTiles.ORDEALS_CASTLE_W, OverworldTiles.ORDEALS_CASTLE_E },
+			new byte[] { OverworldTiles.SARDAS_CAVE },
+			new byte[] { OverworldTiles.TOF_ENTRANCE_W, OverworldTiles.TOF_ENTRANCE_E },
+			new byte[] { OverworldTiles.VOLCANO_TOP_W, OverworldTiles.VOLCANO_TOP_E },
+			new byte[] { OverworldTiles.BAHAMUTS_CAVE },
+			new byte[] { OverworldTiles.MARSH_CAVE }
 		};
 
-		List<byte[]> towns = new List<byte[]>{
-		    new byte[] { OverworldTiles.PRAVOKA },
-		    new byte[] { OverworldTiles.ELFLAND },
-		    new byte[] { OverworldTiles.MELMOND },
-		    new byte[] { OverworldTiles.CRESCENT_LAKE },
-		    new byte[] { OverworldTiles.GAIA },
-		    new byte[] { OverworldTiles.ONRAC },
-		    new byte[] { OverworldTiles.LEFEIN },
+			List<byte[]> towns = new List<byte[]>{
+			new byte[] { OverworldTiles.PRAVOKA },
+			new byte[] { OverworldTiles.ELFLAND },
+			new byte[] { OverworldTiles.MELMOND },
+			new byte[] { OverworldTiles.CRESCENT_LAKE },
+			new byte[] { OverworldTiles.GAIA },
+			new byte[] { OverworldTiles.ONRAC },
+			new byte[] { OverworldTiles.LEFEIN },
 		};
 
-		var candidates = new List<byte[]>(dungeons);
+			var candidates = new List<byte[]>(dungeons);
 
-		if (includeTowns) {
-		    candidates.AddRange(towns);
-		}
+			if (includeTowns) {
+				candidates.AddRange(towns);
+			}
 
-		var tileset = new TileSet(this._rom, TileSet.OverworldIndex);
-		tileset.LoadData();
+			var tileset = new TileSet(this._rom, TileSet.OverworldIndex);
+			tileset.LoadData();
 
-		// clear existing
-		for (int i = 0; i < tileset.TileProperties.Count; i++) {
-		    var tp = tileset.TileProperties[i];
-		    if ((tp.TilePropFunc & TilePropFunc.OWTP_SPEC_MASK) == TilePropFunc.OWTP_SPEC_CHIME) {
-			tp.TilePropFunc &= ~TilePropFunc.OWTP_SPEC_CHIME;
-			tileset.TileProperties[i] = tp;
-		    }
-		}
+			// clear existing
+			for (int i = 0; i < tileset.TileProperties.Count; i++) {
+				var tp = tileset.TileProperties[i];
+				if ((tp.TilePropFunc & TilePropFunc.OWTP_SPEC_MASK) == TilePropFunc.OWTP_SPEC_CHIME) {
+					tp.TilePropFunc &= ~TilePropFunc.OWTP_SPEC_CHIME;
+					tileset.TileProperties[i] = tp;
+				}
+			}
 
-		var chime = candidates.SpliceRandom(rng);
+			var chime = candidates.SpliceRandom(rng);
 
-		foreach (var i in chime) {
-		    var tp = tileset.TileProperties[i];
-		    tp.TilePropFunc |= TilePropFunc.OWTP_SPEC_CHIME;
-		    tileset.TileProperties[i] = tp;
+			foreach (var i in chime) {
+				var tp = tileset.TileProperties[i];
+				tp.TilePropFunc |= TilePropFunc.OWTP_SPEC_CHIME;
+				tileset.TileProperties[i] = tp;
+			}
+
+			tileset.StoreData();
 		}
 
-		tileset.StoreData();
-	    }
+		public void RenderMap()

```

Comment:
I don't know if you were looking at the code in `FF1Randomizer/FF1R/Commands/Progen.cs:RenderMap()` -- I see here you are not producing full-size maps so it wouldn't be exactly the same.  Would it make sense to render it full size and then shrink it down to 1024x1024 or whatever the target size is?  It looks like you are turning all the 16x16 tiles into 4x4 tiles and then rendering from that.

Another idea would be to deliberately render maps that looks like the in-game map.

Reply:
Most image operations are stupidly slow in the browser, so i tuned it to a bearable level. Thus I copied your code and modified it to increase the performance in the browser to reasonable levels. Things that can'T be done in the browser:
- encoding the image into anything other than bmp takes 40+ seconds(just the encoding part).
- drawing images in the transform takes 20+ seconds for all tiles..

Don't get me wrong, the way you did it is the preferable and cleaner way, but it unfortunately doesn't work in the browser. Basically I tried around a bit and that was the only thing I could get to work with reasonable performance.

I can increase the size, but the bmp is already 3MB in size, increasing it to 2048x2048 would make it 12MB. It would take roughly 6 instead of 2 seconds to render. The javascript engine takes a couple of seconds extra to download12MB in base64, but the FileSize would still be a bit ridiculous.